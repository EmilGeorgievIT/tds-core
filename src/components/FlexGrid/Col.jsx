import React from 'react'
import PropTypes from 'prop-types'
import { Subscriber } from 'react-broadcast'

import { Col as ReactFlexboxGridCol } from 'react-flexbox-grid'

import safeRest from '../../utils/safeRest'

const removeProps = ({
  xs,
  sm,
  md,
  lg,
  xl,
  xsOffset,
  smOffset,
  mdOffset,
  lgOffset,
  xlOffset,
  ...rest
}) => safeRest(rest)

const Col = ({ span, offset, children, ...rest }) => (
  <Subscriber channel="gutterless">
    {gutterless => {
      return (
        <ReactFlexboxGridCol
          {...removeProps(rest)}
          xs={span || true}
          xsOffset={offset}
          style={gutterless}
        >
          {children}
        </ReactFlexboxGridCol>
      )
    }}
  </Subscriber>
)

Col.propTypes = {
  span: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
  offset: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]),
  children: PropTypes.node.isRequired,
}

Col.defaultProps = {
  span: undefined,
  offset: undefined,
}

export default Col
