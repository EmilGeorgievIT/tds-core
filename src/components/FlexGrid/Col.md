Columns automatically expand to fill the width of its container. You can specify a span amount from 1 to 12.
