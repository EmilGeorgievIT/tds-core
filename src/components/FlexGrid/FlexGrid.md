The `FlexGrid` component is a wrapper over [react-flexbox-grid](https://github.com/roylee0704/react-flexbox-grid). 

### Basic usage

FlexGrid is handy for basic grid layouts or 1-dimensional positioning. It is used
in conjunction with the [Col](#col) and [Row](#row) components.

```jsx
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col>half</FlexGrid.Col>
    <FlexGrid.Col>half</FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### Responsive example

```jsx
<FlexGrid>
  <Responsive minWidth="sm">
  {
    matches => {
      var desktopSpan = matches ? undefined : 12
      var text = matches ? 'half width on desktop' : 'full width on mobile'

      return (
        <FlexGrid.Row>
          <FlexGrid.Col span={desktopSpan}>{text}</FlexGrid.Col>
          <FlexGrid.Col span={desktopSpan}>{text}</FlexGrid.Col>
        </FlexGrid.Row>
      )
    }
  }
  </Responsive>
</FlexGrid>
```
