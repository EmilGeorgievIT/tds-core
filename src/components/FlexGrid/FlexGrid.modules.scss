@import '../../scss/settings/responsive-grid';
@import '../../scss/utility/responsive';

$max-widths: (xs: auto, small: 544px, medium: 736px, large: 960px, xl: 1168px);

.flexGrid {
  display: flex;
  flex-flow: row wrap;
  padding: 0;
}

.centre {
  margin: 0 auto;
}

/*
  Possible alternatives to responsive CSS
  1) use mutiple 'react-media' queries when it becomes available
  https://github.com/ReactTraining/react-media/issues/69
  2) use multiple Responsive components nested in React 16.2 fragments once all teams
  adopt it https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html
*/
.limitWidth {
  @each $viewport, $max-width in $max-widths {
    @include from-breakpoint($viewport) {
      max-width: $max-width;
      margin: 0 auto;
    }
  }
}

.gutterless {
  padding-left: 0 !important;
  padding-right: 0 !important;
}

.flexGridColumn {
  padding-left: 1rem;
  padding-right: 1rem;
}

// Row styles
.flexGridRow {
  margin: 0;
  flex: 1 1 auto;
}
