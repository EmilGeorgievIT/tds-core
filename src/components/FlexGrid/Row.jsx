import React from 'react'
import PropTypes from 'prop-types'
import { Subscriber } from 'react-broadcast'

import { Row as ReactFlexboxGridRow } from 'react-flexbox-grid'

import safeRest from '../../utils/safeRest'

const Row = ({ children, ...rest }) => (
  <Subscriber channel="flex-grid">
    {gutterStyle => (
      <ReactFlexboxGridRow {...safeRest(rest)} style={gutterStyle}>
        {children}
      </ReactFlexboxGridRow>
    )}
  </Subscriber>
)

Row.propTypes = {
  children: PropTypes.node.isRequired,
}

export default Row
