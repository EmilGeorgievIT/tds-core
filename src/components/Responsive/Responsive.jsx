import React from 'react'
import PropTypes from 'prop-types'
import MediaQuery from 'react-responsive'

import safeRest from '../../utils/safeRest'
import { warn } from '../../utils/warn'

const breakpoints = {
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200,
}

/**
 * Define content within viewport breakpoints
 *
 * <span class="docs--badge__new">new!</span> <span class="docs--badge__version">v0.30.0</span>
 */
const Responsive = ({ minWidth, maxWidth, children, ...rest }) => {
  if (!minWidth && !maxWidth) {
    warn('Responsive', 'Responsive needs a minWidth or maxWith prop')
  }

  const props = {
    ...safeRest(rest),
    minWidth: minWidth ? breakpoints[minWidth] : undefined,
    maxWidth: maxWidth ? breakpoints[maxWidth] - 1 : undefined,
  }

  return <MediaQuery {...props}>{children}</MediaQuery>
}

Responsive.propTypes = {
  /**
   * Set browser min-width media query.
   */
  minWidth: PropTypes.oneOf(['sm', 'md', 'lg', 'xl']),
  /**
   * Set browser max-width media query.
   */
  maxWidth: PropTypes.oneOf(['sm', 'md', 'lg', 'xl']),
  /**
   * The content. Can be text, any HTML element, or any component.
   */
  children: PropTypes.node.isRequired,
}

Responsive.defaultProps = {
  minWidth: undefined,
  maxWidth: undefined,
}

export default Responsive
