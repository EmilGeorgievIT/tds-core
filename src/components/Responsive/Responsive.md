The Responsive component is a thin wrapper over the [**react-media**](https://github.com/ReactTraining/react-media) community 
component. We encourage mobile-first development by using the `minWidth` prop exclusively, though you're welcome to use 
`maxWidth` or any of the other props supplied by **react-media** if the need arises; **note** if you do use props 
supplied by react-media, you must provide pixel values and not breakpoint keys as used by the Responsive component.

### Breakpoints

Using the `minWidth` and `maxWidth` props, you can pass a breakpoint key. They represent the following:

| Breakpoint key | Value as `minWidth` | Value as `maxWidth` |
| -------------- | ------------------- | ------------------- |
| `sm` | 576px | 575px |
| `md` | 768px | 767px |
| `lg` | 992px | 991px |
| `xl` | 1200px | 1199px |

### Basic usage

```jsx
<div>
  <Responsive minWidth="sm" maxWidth="md">
    <Heading level="h1">Hello I'm a heading in XS and SM</Heading>
  </Responsive>
  <Responsive minWidth="md" maxWidth="lg">
    <Card>
      Hello I'm a Card in MD
    </Card>
  </Responsive>
  <Responsive minWidth="lg">
    <Button>Hello I'm a Button in L and XL</Button>
  </Responsive>
</div>
```

### Passing a child function

Since the Responsive component wraps **react-media**, you may pass a function as the only child to perform logical 
operations based on whether or not the query matches the browser viewport.

```jsx
<Responsive minWidth="sm">
{
  (matches) => 
    matches ? "You're using a large display." : "You're viewing this on mobile."
}
</Responsive>
```
