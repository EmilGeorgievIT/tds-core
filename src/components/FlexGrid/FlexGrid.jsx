import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast } from 'react-broadcast'

import { Grid } from 'react-flexbox-grid'

import safeRest from '../../utils/safeRest'

import Col from './Col'
import Row from './Row'

const FlexGrid = ({ gutterless, children, ...rest }) => {
  const css = gutterless ? { padding: 0, margin: 0 } : undefined

  return (
    <Broadcast channel="gutterless" value={css}>
      <Grid {...safeRest(rest)} fluid style={css}>
        {children}
      </Grid>
    </Broadcast>
  )
}

FlexGrid.propTypes = {
  gutterless: PropTypes.bool,
  children: PropTypes.node.isRequired,
}

FlexGrid.defaultProps = {
  gutterless: false,
}

FlexGrid.Row = Row
FlexGrid.Col = Col

export default FlexGrid
