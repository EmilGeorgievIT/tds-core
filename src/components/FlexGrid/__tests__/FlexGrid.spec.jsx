import React from 'react'
import { mount } from 'enzyme'
import { Grid } from 'react-flexbox-grid'

import Responsive from '../../Responsive/Responsive'
import FlexGrid from '../FlexGrid'

describe('FlexGrid', () => {
  const doMount = (props = {}) => {
    const wrapper = mount(<FlexGrid {...props}>Some grid stuff</FlexGrid>)
    return wrapper.find('Grid')
  }

  const doMountWithChildren = (props = {}) => {
    const flexGrid = mount(
      <FlexGrid {...props}>
        <Responsive id="responsive" minWidth="md">
          {match => {
            return (
              <FlexGrid.Col>
                <div>HALF: {match}</div>
              </FlexGrid.Col>
            )
          }}
        </Responsive>
        <FlexGrid.Row>
          <FlexGrid.Col>HALF</FlexGrid.Col>
          <FlexGrid.Col>HALF</FlexGrid.Col>
        </FlexGrid.Row>
      </FlexGrid>
    )

    return {
      findColum: index => flexGrid.find('Col').at(index),
      findRow: index => flexGrid.find('Row').at(index),
      findNowRelated: el => flexGrid.find(el),
      gutterless: { padding: 0, margin: 0 },
    }
  }
  it('renders', () => {
    const flexGrid = doMount()

    expect(flexGrid).toMatchSnapshot()
  })

  it('renders a react-flexbox-grid Grid component', () => {
    const flexGrid = doMount()

    expect(flexGrid).toMatchSelector(Grid)
  })

  it('is fluid by default', () => {
    const flexGrid = doMount()

    expect(flexGrid).toHaveProp('fluid', true)
  })

  it('should render all children related to Grid with no gutter', () => {
    const { findColum, findRow, findNowRelated, gutterless } = doMountWithChildren({
      gutterless: true,
    })

    const col1 = findColum(1)
    const col2 = findColum(3)
    const col3 = findColum(5)
    const row = findRow(1)
    const NonRelatedElement = findNowRelated('Responsive')

    expect(col1).toHaveProp('style', gutterless)
    expect(col2).toHaveProp('style', gutterless)
    expect(col3).toHaveProp('style', gutterless)
    expect(row).toHaveProp('style', gutterless)
    expect(NonRelatedElement).not.toHaveProp('style', gutterless)
  })

  it('passes additional attributes to the element', () => {
    const flexGrid = doMount({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(flexGrid).toHaveProp('id', 'the-id')
    expect(flexGrid).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const flexGrid = doMount({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(flexGrid).not.toHaveProp('className', 'my-custom-class')
    expect(flexGrid).not.toHaveProp('style', { color: 'hotpink' })
  })
})
