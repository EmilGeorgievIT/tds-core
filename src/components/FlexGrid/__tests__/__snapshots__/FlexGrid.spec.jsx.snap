// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexGrid renders 1`] = `
<div
  class="flexGrid container-fluid"
>
  <div
    class="flexGridRow row"
  >
    <div
      class="col-xs col-xs-offset flexGridColumn"
    >
      1st column content
    </div>
    <div
      class="col-xs col-xs-offset flexGridColumn"
    >
      2nd column content
    </div>
  </div>
</div>
`;
